### PR TITLE
Support grounding dino

### PR DESCRIPTION
**Pull request by YanxingLiu**
# Grounding DINO: Marrying DINO with Grounded Pre-Training for Open-Set Object Detection

[GLIP: Grounded Language-Image Pre-training](https://arxiv.org/abs/2112.03857)

## Abstract

In this paper, we present an open-set object detector, called Grounding DINO, by marrying Transformer-based detector DINO with grounded pre-training, which can detect arbitrary objects with human inputs such as category names or referring expressions. The key solution of open-set object detection is introducing language to a closed-set detector for open-set concept generalization. To effectively fuse language and vision modalities, we conceptually divide a closed-set detector into three phases and propose a tight fusion solution, which includes a feature enhancer, a language-guided query selection, and a cross-modality decoder for cross-modality fusion. While previous works mainly evaluate open-set object detection on novel categories, we propose to also perform evaluations on referring expression comprehension for objects specified with attributes. Grounding DINO performs remarkably well on all three settings, including benchmarks on COCO, LVIS, ODinW, and RefCOCO/+/g. Grounding DINO achieves a 52.5 AP on the COCO detection zero-shot transfer benchmark, i.e., without any training data from COCO. It sets a new record on the ODinW zero-shot benchmark with a mean 26.1 AP.

<div align=center>

</div>

## Installation

```shell
cd $MMDETROOT

# source installation
pip install -r requirements/multimodal.txt

# or mim installation
mim install mmdet[multimodal]
```

```
cd $MMDETROOT

wget https://github.com/IDEA-Research/GroundingDINO/releases/download/v0.1.0-alpha/groundingdino_swint_ogc.pth

python projects/GroundingDINO/tools/model_converters/groundingdino_to_mmdet.py \
		groundingdino_swint_ogc.pth \
		weights/groundingdino_swint_ogc_mmdet.pth
# this script will generate a model $WEIGHT_FILE in $MMDETROOT

python demo/image_demo.py \
	demo/demo.jpg \
	projects/GroundingDINO/configs/groundingdino/groundingdino_swin-t.py \
	--weights  $WEIGHT_FILE \
	--texts 'bench . car .'
```

<div align=center>

</div>

## Results and Models

|      Model       | backbone | COCO mAP |                  Pre-Train Data                  |                         Config                          |                                                        Download                                                         |
| :--------------: | :------: | :------: | :----------------------------------------------: | :-----------------------------------------------------: | :---------------------------------------------------------------------------------------------------------------------: |
| Grounding DINO-T |  Swin-T  |   48.5   |                 O365,GoldG,Cap4M                 | [config](configs/groundingdino/groundingdino_swin-t.py) |   [model](https://github.com/IDEA-Research/GroundingDINO/releases/download/v0.1.0-alpha/groundingdino_swint_ogc.pth)    |
| Grounding DINO-B |  Swin-B  |   56.9   | COCO,O365,GoldG,Cap4M,OpenImage,ODinW-35,RefCOCO | [config](configs/groundingdino/groundingdino_swin-b.py) | [model](https://github.com/IDEA-Research/GroundingDINO/releases/download/v0.1.0-alpha2/groundingdino_swinb_cogcoor.pth) |

Note:

1. The weights corresponding to the zero-shot model are adopted from the official weights and converted using the [script](./tools/model_converters/groundingdino_to_mmdet.py). We have not retrained the model for the time being.

### CODE REVIEW

File: projects/GroundingDINO/README.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
+# Grounding DINO: Marrying DINO with Grounded Pre-Training for Open-Set Object Detection
```
Comment (hhaAndroid) [inline projects/GroundingDINO/README.md:1]: 可以先放这里。等做的差不多，就还是放到 package 里面去

File: projects/GroundingDINO/configs/groundingdino/groundingdino_swin-t.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
+_base_ = [
```
Comment (hhaAndroid) [inline projects/GroundingDINO/configs/groundingdino/groundingdino_swin-t.py:1]: 文件的命名参考 glip 的写法比较合适

File: projects/GroundingDINO/configs/groundingdino/groundingdino_swin-t.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    type='GroundingDINO',
+    num_queries=900,
+    with_box_refine=True,
+    as_two_stage=True,
+    data_preprocessor=dict(
+        type='DetDataPreprocessor',
+        mean=[123.675, 116.28, 103.53],
+        std=[58.395, 57.12, 57.375],
+        bgr_to_rgb=True,
+        pad_mask=True,
```
Comment (hhaAndroid) [inline projects/GroundingDINO/configs/groundingdino/groundingdino_swin-t.py:21]: 这个应该不需要。因为这是纯检测算法

File: projects/GroundingDINO/configs/groundingdino/groundingdino_swin-t.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    ),
+    language_model_cfg=dict(
+        type='BertModelGroundingDINO',
+        name='bert-base-uncased',
```
Comment (hhaAndroid) [inline projects/GroundingDINO/configs/groundingdino/groundingdino_swin-t.py:25]: 这个参数和 GLIP 一样，放到最前面去吧

File: projects/GroundingDINO/configs/groundingdino/groundingdino_swin-t.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        bias=True,
+        norm_cfg=dict(type='GN', num_groups=32),
+        num_outs=4),
+    encoder=dict(
+        num_layers=6,
+        layer_cfg=dict(
+            self_attn_cfg=dict(embed_dims=256, num_levels=4, dropout=0.0),
+            ffn_cfg=dict(
+                embed_dims=256, feedforward_channels=2048,
+                ffn_drop=0.0)),  # visual layer config
```
Comment (hhaAndroid) [inline projects/GroundingDINO/configs/groundingdino/groundingdino_swin-t.py:62]: visual layer config 注释应该放到 58 行后，这样比较容易看

File: projects/GroundingDINO/groundingdino/detectors/grounding_dino.py [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+from mmdet.registry import MODELS
+from mmdet.structures import OptSampleList
+from ..layers.grounding_dino_layers import (GroundingDinoTransformerDecoder,
+                                            GroundingDinoTransformerEncoder)
+
+
+@MODELS.register_module()
+class GroundingDINO(DINO):
+
+    def __init__(self, language_model_cfg, *args, **kwargs) -> None:
```
Comment (hhaAndroid) [inline projects/GroundingDINO/groundingdino/detectors/grounding_dino.py:23]: 输入参数命名尽量和 GLIP 一致，这样用户比较好用。

File: projects/GroundingDINO/groundingdino/detectors/grounding_dino.py [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+        head_inputs_dict = dict(
+            enc_outputs_class=topk_score,
+            enc_outputs_coord=topk_coords,
+            dn_meta=dn_meta) if self.training else dict()
+        # append text_feats to head_inputs_dict
+        head_inputs_dict['memory_text'] = memory_text
+        head_inputs_dict['text_token_mask'] = text_token_mask
+        return decoder_inputs_dict, head_inputs_dict
+
+    def get_tokens_positive_and_prompts(
```
Comment (hhaAndroid) [inline projects/GroundingDINO/groundingdino/detectors/grounding_dino.py:174]: 这个地方 glip 有更新，你这边也参考最新的写法同步下

File: projects/GroundingDINO/groundingdino/detectors/grounding_dino.py [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+                *_positive_maps_and_prompts)
+
+            self._text_dict = self.language_model(text_prompts)
+            if self.feat_map is not None:
+                self._text_dict['embedded'] = self.feat_map(
+                    self._text_dict['embedded'])
+
+    def predict(self, batch_inputs, batch_data_samples, rescale: bool = True):
+        # text feature extraction
+        self.extract_text_feat(batch_inputs, batch_data_samples)
```
Comment (hhaAndroid) [inline projects/GroundingDINO/groundingdino/detectors/grounding_dino.py:257]: 这个函数太大了，内部其实做了很多事情，直接叫做 self.extract_text_feat 应该不妥

File: projects/GroundingDINO/groundingdino/layers/fuse_modules.py [hunk 9]
```diff
@@ -0,0 +1,394 @@
...
+        attn_output_l = attn_output_l.transpose(1, 2)
+        attn_output_l = attn_output_l.reshape(bsz, src_len, self.embed_dim)
+
+        attn_output_v = self.out_v_proj(attn_output_v)
+        attn_output_l = self.out_l_proj(attn_output_l)
+
+        return attn_output_v, attn_output_l
+
+
+class BiMultiHeadAttention(nn.Module):
```
Comment (hhaAndroid) [inline projects/GroundingDINO/groundingdino/layers/fuse_modules.py:170]: 这个和 GLIP 里面是一样的吗

File: projects/GroundingDINO/groundingdino/layers/grounding_dino_layers.py [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+from mmdet.utils import ConfigType, OptConfigType
+from ..utils import get_sine_pos_embed
+from .fuse_modules import BiAttentionBlock
+
+
+class GroundingDinoTransformerDecoderLayer(
+        DeformableDetrTransformerDecoderLayer):
+
+    def __init__(self,
+                 ca_text_cfg: OptConfigType = dict(
```
Comment (hhaAndroid) [inline projects/GroundingDINO/groundingdino/layers/grounding_dino_layers.py:25]: ca 的简称不对，应该是cross_attn_

File: projects/GroundingDINO/configs/groundingdino/groundingdino_swin-t.py [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+        fusion_layer_cfg=dict(
+            v_dim=256, l_dim=256, embed_dim=1024,
+            num_heads=4),  # fusion layer config
+    ),
+    decoder=dict(
+        num_layers=6,
+        return_intermediate=True,
+        layer_cfg=dict(
+            self_attn_cfg=dict(embed_dims=256, num_heads=8, dropout=0.0),
+            ca_text_cfg=dict(embed_dims=256, num_heads=8, dropout=0.0),
```
Comment (hhaAndroid) [inline projects/GroundingDINO/configs/groundingdino/groundingdino_swin-t.py:77]: ca_text_cfg -> cross_attn_v2t_cfg
cross_attn_cfg -> cross_attn_t2v_cfg

File: projects/GroundingDINO/configs/groundingdino/groundingdino_swin-t.py [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+        in_channels=[192, 384, 768],
+        kernel_size=1,
+        out_channels=256,
+        act_cfg=None,
+        bias=True,
+        norm_cfg=dict(type='GN', num_groups=32),
+        num_outs=4),
+    encoder=dict(
+        num_layers=6,
+        layer_cfg=dict(
```
Comment (hhaAndroid) [inline projects/GroundingDINO/configs/groundingdino/groundingdino_swin-t.py:58]: 最好改成 visual_layer_cfg

File: projects/GroundingDINO/configs/groundingdino/groundingdino_swin-t.py (hunk 2 above)
Comment (YanxingLiu) [reply projects/GroundingDINO/configs/groundingdino/groundingdino_swin-t.py]: 已经修改文件名

File: mmdet/models/detectors/grounding_dino.py [hunk 13]
```diff
@@ -0,0 +1,310 @@
...
+    <https://arxiv.org/abs/2303.05499>`_
+
+    Code is modified from the `official github repo
+    <https://github.com/IDEA-Research/GroundingDINO>`_.
+    """
+
+    def __init__(self, language_model, *args, **kwargs) -> None:
+
+        self.language_model_cfg = language_model
+        self._text_prompts = None
```
Comment (hhaAndroid) [inline mmdet/models/detectors/grounding_dino.py:34]: 这些参考我最新的 glip pr 也删掉吧，不用缓存了，会导致一些隐藏的bug

File: mmdet/models/detectors/grounding_dino.py [hunk 14]
```diff
@@ -0,0 +1,310 @@
...
+                    padding='max_length'
+                    if self.language_model.pad_to_max else 'longest',
+                    return_tensors='pt')
+            tokens_positive, noun_phrases = run_ner(original_caption)
+            self._entities = noun_phrases
+            caption_string = original_caption
+
+        return tokenized, caption_string, tokens_positive
+
+    def get_positive_map(self, tokenized, tokens_positive):
```
Comment (hhaAndroid) [inline mmdet/models/detectors/grounding_dino.py:112]: 有些函数如果 glip 里面也有，并且你没有改，那就直接调用就行，不用重写

File: mmdet/models/utils/vlfuse_helper.py [hunk 15]
```diff
@@ -134,10 +134,11 @@ def _reset_parameters(self):
         self.out_l_proj.bias.data.fill_(0)
 
     def forward(
```
Comment (hhaAndroid) [inline mmdet/models/utils/vlfuse_helper.py:136]: atten_mask_v 是不是应该放最后，这样 glip 调用的时候应该会报错的

File: mmdet/models/utils/vlfuse_helper.py [hunk 16]
```diff
@@ -292,39 +300,55 @@ def forward(
...
 
     def single_attention_call(
-            self,
-            visual: Tensor,
-            lang: Tensor,
-            attention_mask_l: Optional[Tensor] = None
+        self,
+        visual: Tensor,
+        lang: Tensor,
+        attention_mask_v: Optional[Tensor] = None,
```
Comment (hhaAndroid) [inline mmdet/models/utils/vlfuse_helper.py:350]: 这里也是一样

File: mmdet/models/utils/vlfuse_helper.py (hunk 15 above)
Comment (YanxingLiu) [reply mmdet/models/utils/vlfuse_helper.py:136]: glip 我也稍微调了下顺序，跑的时候可以正常运行的，因为感觉 特征用 v l的顺序，mask 用lv的顺序有点别扭。

File: configs/grounding_dino/README.md [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+wget https://github.com/IDEA-Research/GroundingDINO/releases/download/v0.1.0-alpha/groundingdino_swint_ogc.pth
+
+python tools/model_converters/groundingdino_to_mmdet.py \
+		groundingdino_swint_ogc.pth \
+		weights/groundingdino_swint_ogc_mmdet.pth
+# this script will generate a model $WEIGHT_FILE in $MMDETROOT
+
+python demo/image_demo.py \
+	demo/demo.jpg \
+	projects/GroundingDINO/configs/groundingdino/groundingdino_swin-t.py \
```
[the remark this answers is not in the document]
Comment (YanxingLiu) [reply configs/grounding_dino/README.md:39]: 已经改掉了

File: configs/grounding_dino/README.md [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+
+<div align=center>
+<img src="https://github.com/open-mmlab/mmdetection/assets/42299757/3a3bd6f1-e2ed-43d4-aa22-0bb07ee6f20b"/>
+</div>
+
+## Results and Models
+
+|      Model       | backbone | COCO mAP |                  Pre-Train Data                  |                                              Config                                              |                                                        Download                                                        |
+| :--------------: | :------: | :------: | :----------------------------------------------: | :----------------------------------------------------------------------------------------------: | :--------------------------------------------------------------------------------------------------------------------: |
+| Grounding DINO-T |  Swin-T  |   48.5   |                 O365,GoldG,Cap4M                 |                      [config](grounding_dino_swin-b_pretrain_mixeddata.py)                       |   [model](https://download.openmmlab.com/mmdetection/v3.0/grounding_dino/groundingdino_swint_ogc_mmdet-822d7e9d.pth)   |
```
Comment (hhaAndroid) [inline configs/grounding_dino/README.md:47]: config 配置写错了

File: configs/grounding_dino/README.md [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+| Grounding DINO-B |  Swin-B  |   56.9   | COCO,O365,GoldG,Cap4M,OpenImage,ODinW-35,RefCOCO | [config](grounding_dino_swin-b_pretrain_mixeddata.pygrounding_dino_swin-b_pretrain_mixeddata.py) | [model](https://download.openmmlab.com/mmdetection/v3.0/grounding_dino/groundingdino_swinb_cogcoor_mmdet-55949c9c.pth) |
```
Comment (hhaAndroid) [inline configs/grounding_dino/README.md:48]: 这个也写错了

File: mmdet/models/detectors/grounding_dino.py [hunk 20]
```diff
@@ -0,0 +1,309 @@
...
+                    filter(lambda x: len(x) > 0, original_caption))
+
+            caption_string = ''
+            tokens_positive = []
+            for idx, word in enumerate(original_caption):
+                tokens_positive.append(
+                    [[len(caption_string),
+                      len(caption_string) + len(word)]])
+                caption_string += word
+                caption_string += self._special_tokens
```
Comment (hhaAndroid) [inline mmdet/models/detectors/grounding_dino.py:84]: 最后一个也要加 self._special_tokens 吗?

File: mmdet/models/detectors/grounding_dino.py [hunk 21]
```diff
@@ -0,0 +1,309 @@
...
+            entities = original_caption
+        else:
+            if original_caption.endswith(self._special_tokens):
+                original_caption = original_caption.replace(
+                    self._special_tokens, '')
+                # NOTE: Tokenizer in Grounding DINO is different from
+                # that in GLIP. The tokenizer in GLIP will pad the
+                # caption_string to max_length, while the tokenizer
+                # in Grounding DINO will not.
+                tokenized = self.language_model.tokenizer(
```
Comment (hhaAndroid) [inline mmdet/models/detectors/grounding_dino.py:103]: 这个分支就不需要吗?

File: mmdet/models/detectors/grounding_dino.py [hunk 22]
```diff
@@ -0,0 +1,309 @@
...
+        else:
+            _positive_maps_and_prompts = [
+                self.get_tokens_positive_and_prompts(text_prompt,
+                                                     custom_entities)
+                for text_prompt in text_prompts
+            ]
+        token_positive_maps, text_prompts, _, entities = zip(
+            *_positive_maps_and_prompts)
+        # extract text feats
+        text_dict = self.language_model(text_prompts)
```
Comment (hhaAndroid) [inline mmdet/models/detectors/grounding_dino.py:275]: 你看下最新的 glip 分支,这里必须要一个 list的

File: mmdet/models/detectors/grounding_dino.py [hunk 23]
```diff
@@ -0,0 +1,309 @@
...
+        # image feature extraction
+        visual_feats = self.extract_feat(batch_inputs)
+
+        head_inputs_dict = self.forward_transformer(visual_feats, text_dict,
+                                                    batch_data_samples)
+        results_list = self.bbox_head.predict(
+            **head_inputs_dict,
+            rescale=rescale,
+            batch_data_samples=batch_data_samples)
+        for data_sample, pred_instances in zip(batch_data_samples,
```
Comment (hhaAndroid) [inline mmdet/models/detectors/grounding_dino.py:292]: entities 要拆解,你看下最新的 glip eval 代码

File: mmdet/models/detectors/grounding_dino.py (hunk 20 above)
Comment (YanxingLiu) [reply mmdet/models/detectors/grounding_dino.py:84]: > 最后一个也要加 self._special_tokens 吗?

是的，不然精度会有0.5个点的下降

File: mmdet/models/detectors/grounding_dino.py [hunk 24]
```diff
@@ -0,0 +1,309 @@
...
+            # caption_string to max_length, while the tokenizer
+            # in Grounding DINO will not.
+            tokenized = self.language_model.tokenizer(
+                [caption_string],
+                padding='max_length'
+                if self.language_model.pad_to_max else 'longest',
+                return_tensors='pt')
+            entities = original_caption
+        else:
+            if original_caption.endswith(self._special_tokens):
```
Comment (hhaAndroid) [inline mmdet/models/detectors/grounding_dino.py:96]: 这个逻辑有点问题, 如果 if 分支没有进就啥都没有了

File: mmdet/models/utils/vlfuse_helper.py [hunk 25]
```diff
@@ -388,12 +415,15 @@ def forward(self, x: dict) -> dict:
 
         if self.use_checkpoint:
             fused_visual_features, language_features = checkpoint.checkpoint(
-                self.b_attn, visual_features, language_dict_features['hidden'],
-                language_dict_features['masks'])
+                self.b_attn,
```
Comment (hhaAndroid) [inline mmdet/models/utils/vlfuse_helper.py:418]: 这个地方有问题,你没有同步最新的代码,这个地方应该改了

**Review by hhaAndroid — APPROVED** [review]

**Comment by HenryPeking** [comment]
Is the grouding DINO support finetune?

**Comment by YanxingLiu** [comment]
We did not test the training phase as the original code was not open for training related content. If you want to try to fine-tune it, you may need to modify the configuration file a bit. Thank you for your interest.

**Comment by SoulProficiency** [comment]
What are the minimum equipment requirements of fine-tunning ground DINO with coco dataset？（FP32&total parameters&batch-size≥32）

**Pull request merged**